`transpose` panics on `null` input

### Terraform Version

```shell
Terraform v1.10.5
on linux_amd64
```

### Terraform Configuration Files

```terraform
output "output_name" {
  value = transpose({ "test" = null })
}
```


### Debug Output

```
2025-02-20T16:16:24.710+0100 [INFO]  Terraform version: 1.10.5
2025-02-20T16:16:24.710+0100 [DEBUG] using github.com/hashicorp/go-tfe v1.70.0
2025-02-20T16:16:24.710+0100 [DEBUG] using github.com/hashicorp/hcl/v2 v2.23.0
2025-02-20T16:16:24.710+0100 [DEBUG] using github.com/hashicorp/terraform-svchost v0.1.1
2025-02-20T16:16:24.710+0100 [DEBUG] using github.com/zclconf/go-cty v1.16.2
2025-02-20T16:16:24.710+0100 [INFO]  Go runtime version: go1.23.3
2025-02-20T16:16:24.710+0100 [INFO]  CLI args: []string{"terraform", "plan", "-refresh=false"}
2025-02-20T16:16:24.710+0100 [DEBUG] Attempting to open CLI config file: /home/bloodyfool/.terraformrc
2025-02-20T16:16:24.710+0100 [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2025-02-20T16:16:24.710+0100 [INFO]  Loading CLI configuration from /home/bloodyfool/.terraform.d/credentials.tfrc.json
2025-02-20T16:16:24.710+0100 [DEBUG] ignoring non-existing provider search directory terraform.d/plugins
2025-02-20T16:16:24.710+0100 [DEBUG] ignoring non-existing provider search directory /home/bloodyfool/.terraform.d/plugins
2025-02-20T16:16:24.710+0100 [DEBUG] ignoring non-existing provider search directory /home/bloodyfool/.local/share/terraform/plugins
2025-02-20T16:16:24.710+0100 [DEBUG] ignoring non-existing provider search directory /usr/share/i3-gnome/terraform/plugins
2025-02-20T16:16:24.710+0100 [DEBUG] ignoring non-existing provider search directory /usr/share/gnome/terraform/plugins
2025-02-20T16:16:24.710+0100 [DEBUG] ignoring non-existing provider search directory /usr/local/share/terraform/plugins
2025-02-20T16:16:24.710+0100 [DEBUG] ignoring non-existing provider search directory /usr/share/terraform/plugins
2025-02-20T16:16:24.710+0100 [DEBUG] ignoring non-existing provider search directory /var/lib/snapd/desktop/terraform/plugins
2025-02-20T16:16:24.710+0100 [DEBUG] ignoring non-existing provider search directory /home/bloodyfool/.nix-profile/share/terraform/plugins
2025-02-20T16:16:24.710+0100 [DEBUG] ignoring non-existing provider search directory /nix/var/nix/profiles/default/share/terraform/plugins
2025-02-20T16:16:24.710+0100 [DEBUG] ignoring non-existing provider search directory /home/bloodyfool/.nix-profile/share/terraform/plugins
2025-02-20T16:16:24.710+0100 [DEBUG] ignoring non-existing provider search directory /nix/var/nix/profiles/default/share/terraform/plugins
2025-02-20T16:16:24.710+0100 [INFO]  CLI command args: []string{"plan", "-refresh=false"}
2025-02-20T16:16:24.711+0100 [DEBUG] checking for provisioner in "."
2025-02-20T16:16:24.711+0100 [DEBUG] checking for provisioner in "/home/bloodyfool/bin"
2025-02-20T16:16:24.711+0100 [INFO]  backend/local: starting Plan operation
2025-02-20T16:16:24.711+0100 [DEBUG] Building and walking validate graph
2025-02-20T16:16:24.711+0100 [DEBUG] ReferenceTransformer: "output.output_name (expand)" references: []
2025-02-20T16:16:24.711+0100 [DEBUG] Starting graph walk: walkValidate
2025-02-20T16:16:24.712+0100 [ERROR] vertex "output.output_name" error: Error in function call
2025-02-20T16:16:24.712+0100 [ERROR] vertex "output.output_name (expand)" error: Error in function call
╷
│ Error: Error in function call
│
│   on main.tf line 2, in output "output_name":
│    2:   value = transpose({ "test" = null })
│     ├────────────────
│     │ while calling transpose(values)
│
│ Call to function "transpose" failed: panic in function implementation: can't use ElementIterator on null value
│ goroutine 72 [running]:
│ runtime/debug.Stack()
│ 	runtime/debug/stack.go:26 +0x5e
│ github.com/zclconf/go-cty/cty/function.errorForPanic(...)
│ 	github.com/zclconf/go-cty@v1.16.2/cty/function/error.go:44
│ github.com/zclconf/go-cty/cty/function.Function.Call.func2()
│ 	github.com/zclconf/go-cty@v1.16.2/cty/function/function.go:349 +0x9b
│ panic({0x3050ca0?, 0x3ddb130?})
│ 	runtime/panic.go:785 +0x132
│ github.com/zclconf/go-cty/cty.Value.ElementIterator({{{0x3e21418?, 0xc0007f76f0?}}, {0x0?, 0x0?}})
│ 	github.com/zclconf/go-cty@v1.16.2/cty/value_ops.go:1219 +0xbe
│ github.com/hashicorp/terraform/internal/lang/funcs.init.func19({0xc0009a19a0?, 0xc0007f7740?, 0x3198f00?}, {{0x3e21248?, 0xc00058fb70?}})
│ 	github.com/hashicorp/terraform/internal/lang/funcs/collection.go:585 +0x20d
│ github.com/zclconf/go-cty/cty/function.Function.Call({0x3e21210?}, {0xc0009a19a0, 0x1, 0x1})
│ 	github.com/zclconf/go-cty@v1.16.2/cty/function/function.go:353 +0x6bf
│ github.com/hashicorp/hcl/v2/hclsyntax.(*FunctionCallExpr).Value(0xc0009c0000, 0xc0007f38a8)
│ 	github.com/hashicorp/hcl/v2@v2.23.0/hclsyntax/expression.go:528 +0x1acf
│ github.com/hashicorp/terraform/internal/lang.(*Scope).EvalExpr(0xc0003a4d80, {0x3e21828, 0xc0009c0000}, {{0x3e21918?, 0x5a441a0?}})
│ 	github.com/hashicorp/terraform/internal/lang/eval.go:179 +0x1a5
│ github.com/hashicorp/terraform/internal/terraform.(*BuiltinEvalContext).EvaluateExpr(0x0?, {0x3e21828, 0xc0009c0000}, {{0x3e21918?, 0x5a441a0?}}, {0x0?, 0x0?})
│ 	github.com/hashicorp/terraform/internal/terraform/eval_context_builtin.go:331 +0xab
│ github.com/hashicorp/terraform/internal/terraform.(*NodeApplyableOutput).Execute(0xc0005ea600, {0x3e48f38, 0xc000124700}, 0x50?)
│ 	github.com/hashicorp/terraform/internal/terraform/node_output.go:439 +0x6dc
│ github.com/hashicorp/terraform/internal/terraform.(*ContextGraphWalker).Execute(0xc0009fa000, {0x3e48f38, 0xc000124700}, {0x7c044224bb68, 0xc0005ea600})
│ 	github.com/hashicorp/terraform/internal/terraform/graph_walk_context.go:161 +0xb5
│ github.com/hashicorp/terraform/internal/terraform.(*Graph).walk.func1({0x3534740, 0xc0005ea600})
│ 	github.com/hashicorp/terraform/internal/terraform/graph.go:143 +0x7c3
│ github.com/hashicorp/terraform/internal/dag.(*Walker).walkVertex(0xc0009b9800, {0x3534740, 0xc0005ea600}, 0xc0007faec0)
│ 	github.com/hashicorp/terraform/internal/dag/walk.go:384 +0x2d1
│ created by github.com/hashicorp/terraform/internal/dag.(*Walker).Update in goroutine 70
│ 	github.com/hashicorp/terraform/internal/dag/walk.go:307 +0xfb3
│ .
╵

```


### Expected Behavior

No Panic

### Actual Behavior

Panic

### Steps to Reproduce

`terraform plan`

### Additional Context

_No response_

### References

_No response_

### Generative AI / LLM assisted development?

_No response_

--- a/.changes/v1.11/BUG FIXES-20250303-125722.yaml
+++ b/.changes/v1.11/BUG FIXES-20250303-125722.yaml
@@ -1,0 +1,5 @@
+kind: BUG FIXES
+body: 'lang/funcs/transpose: Avoid crash due to map with null values'
+time: 2025-03-03T12:57:22.400359Z
+custom:
+    Issue: "36611"

--- a/internal/lang/funcs/collection.go
+++ b/internal/lang/funcs/collection.go
@@ -582,8 +582,14 @@ var TransposeFunc = function.New(&function.Spec{
 
 		for it := inputMap.ElementIterator(); it.Next(); {
 			inKey, inVal := it.Element()
+			if inVal.IsNull() {
+				return cty.MapValEmpty(cty.List(cty.String)), errors.New("input must not contain null list")
+			}
 			for iter := inVal.ElementIterator(); iter.Next(); {
 				_, val := iter.Element()
+				if val.IsNull() {
+					return cty.MapValEmpty(cty.List(cty.String)), errors.New("input list must not contain null string")
+				}
 				if !val.Type().Equals(cty.String) {
 					return cty.MapValEmpty(cty.List(cty.String)), errors.New("input must be a map of lists of strings")
 				}

--- a/internal/lang/funcs/collection_test.go
+++ b/internal/lang/funcs/collection_test.go
@@ -1833,6 +1833,37 @@ func TestTranspose(t *testing.T) {
 			}).WithMarks(cty.NewValueMarks("beep", "boop", "bloop")),
 			false,
 		},
+		{
+			cty.NullVal(cty.Map(cty.List(cty.String))),
+			cty.NilVal,
+			true,
+		},
+		{
+			cty.MapVal(map[string]cty.Value{
+				"test": cty.NullVal(cty.List(cty.String)),
+			}),
+			cty.NilVal,
+			true,
+		},
+		{
+			cty.MapVal(map[string]cty.Value{
+				"test": cty.ListVal([]cty.Value{cty.NullVal(cty.String)}),
+			}),
+			cty.NilVal,
+			true,
+		},
+		{
+			cty.UnknownVal(cty.Map(cty.List(cty.String))),
+			cty.UnknownVal(cty.Map(cty.List(cty.String))).RefineNotNull(),
+			false,
+		},
+		{
+			cty.MapVal(map[string]cty.Value{
+				"test": cty.ListVal([]cty.Value{cty.UnknownVal(cty.String)}),
+			}),
+			cty.UnknownVal(cty.Map(cty.List(cty.String))).RefineNotNull(),
+			false,
+		},
 	}
 
 	for _, test := range tests {

